src/tensors/tensor.jl

```diff
@@ -219,101 +239,120 @@ function _buildblockstructure(P::ProductSpace{S,N}, blocksectors) where {S<:Inde
+Create a `TensorMap` with element type `T`, of all zeros with spaces specified by `codomain` and `domain`.
 """
-function TensorMap(f, ::Type{T}, codom::ProductSpace{S},
-                   dom::ProductSpace{S}) where {S<:IndexSpace,T<:Number}
-    return TensorMap(d -> f(T, d), codom, dom)
-end
-
-function TensorMap(::Type{T}, codom::ProductSpace{S},
-                   dom::ProductSpace{S}) where {S<:IndexSpace,T<:Number}
-    return TensorMap(d -> Array{T}(undef, d), codom, dom)
-end
-
-function TensorMap(::UndefInitializer, ::Type{T}, codom::ProductSpace{S},
-                   dom::ProductSpace{S}) where {S<:IndexSpace,T<:Number}
-    return TensorMap(d -> Array{T}(undef, d), codom, dom)
-end
-
-function TensorMap(::UndefInitializer, codom::ProductSpace{S},
-                   dom::ProductSpace{S}) where {S<:IndexSpace}
-    return TensorMap(undef, Float64, codom, dom)
-end
-
-function TensorMap(::Type{T}, codom::TensorSpace{S},
-                   dom::TensorSpace{S}) where {T<:Number,S<:IndexSpace}
-    return TensorMap(T, convert(ProductSpace, codom), convert(ProductSpace, dom))
-end
-
-function TensorMap(dataorf, codom::TensorSpace{S},
-                   dom::TensorSpace{S}) where {S<:IndexSpace}
-    return TensorMap(dataorf, convert(ProductSpace, codom), convert(ProductSpace, dom))
-end
+Base.zeros
 
-function TensorMap(dataorf, ::Type{T}, codom::TensorSpace{S},
-                   dom::TensorSpace{S}) where {T<:Number,S<:IndexSpace}
-    return TensorMap(dataorf, T, convert(ProductSpace, codom), convert(ProductSpace, dom))
-end
+@doc """
+    ones([T=Float64,], codomain::ProductSpace{S,N₁}, domain::ProductSpace{S,N₂}) where {S,N₁,N₂,T}
+    ones([T=Float64,], codomain ← domain)
+    
+Create a `TensorMap` with element type `T`, of all ones with spaces specified by `codomain` and `domain`.
+"""
+Base.ones
 
-function TensorMap(codom::TensorSpace{S}, dom::TensorSpace{S}) where {S<:IndexSpace}
-    return TensorMap(Float64, convert(ProductSpace, codom), convert(ProductSpace, dom))
+for (fname, felt) in ((:zeros, :zero), (:ones, :one))
+    @eval begin
+        function Base.$fname(codomain::TensorSpace{S},
+                             domain::TensorSpace{S}=one(codomain)) where {S<:IndexSpace}
+            return Base.$fname(codomain ← domain)
+        end
+        function Base.$fname(::Type{T}, codomain::TensorSpace{S},
+                             domain::TensorSpace{S}=one(codomain)) where {T,S<:IndexSpace}
+            return Base.$fname(T, codomain ← domain)
+        end
+        Base.$fname(V::TensorMapSpace) = Base.$fname(Float64, V)
+        function Base.$fname(::Type{T}, V::TensorMapSpace) where {T}
+            t = TensorMap{T}(undef, V)
+            fill!(t, $felt(T))
+            return t
+        end
+    end
 end
 
-function TensorMap(dataorf, T::Type{<:Number}, P::TensorMapSpace{S}) where {S<:IndexSpace}
-    return TensorMap(dataorf, T, codomain(P), domain(P))
-end
+for randfun in (:rand, :randn, :randexp)
```

Comment:
`randhaar`/`randisometry` is lost in this process, but that one is anyway slightly different, since it needs to be interpreted at the level of whole blocks rather than elementswise, and as such is also harder to be applied in place.